packages/labs/src/multi-select/multiSelect.tsx

```diff
@@ -0,0 +1,299 @@
+    isOpen?: boolean;
+    query?: string;
+    selectedItems?: T[];
+}
+
+@PureRender
+export class MultiSelect<T> extends AbstractComponent<IMultiSelectProps<T>, IMultiSelectState<T>> {
+    public static displayName = "Blueprint.MultiSelect";
+
+    public static ofType<T>() {
+        return MultiSelect as new () => MultiSelect<T>;
+    }
+
+    public state: IMultiSelectState<T> = {
+        isOpen: false,
+        query: "",
+        selectedItems: [],
+    };
+
+    private TypedQueryList = QueryList.ofType<T>();
+    private input: HTMLInputElement;
+    private queryList: QueryList<T>;
+    private queryListKeyDown: React.EventHandler<React.KeyboardEvent<HTMLElement>>;
+    private refHandlers = {
+        input: (ref: HTMLInputElement) => this.input = ref,
+        queryList: (ref: QueryList<T>) => this.queryList = ref,
+    };
+
+    public componentDidMount() {
+        // can't use defaultProps because "static members
+        // cannot reference class type parameters"
+        this.setState({ selectedItems: this.props.selectedItems || [] });
+    }
+
+    public render() {
+        // omit props specific to this component, spread the rest.
+        const { itemRenderer, noResults, popoverProps, tagInputProps, ...props } = this.props;
+
+        return <this.TypedQueryList
+            {...props}
+            activeItem={this.state.activeItem}
+            onActiveItemChange={this.handleActiveItemChange}
+            onItemSelect={this.handleItemSelect}
+            query={this.state.query}
+            ref={this.refHandlers.queryList}
+            renderer={this.renderQueryList}
+        />;
+    }
+
+    private renderQueryList = (listProps: IQueryListRendererProps<T>) => {
+        const { tagInputProps = {}, popoverProps = {} } = this.props;
+        const { handleKeyDown, handleKeyUp, query } = listProps;
+        const defaultInputProps: HTMLInputProps = {
+            onChange: this.handleQueryChange,
+            placeholder: "Search...",
+            ref: this.refHandlers.input,
+            value: query,
+        };
+
+        this.queryListKeyDown = handleKeyDown;
+
+        return (
+            <Popover
+                autoFocus={false}
+                canEscapeKeyClose={true}
+                enforceFocus={false}
+                isOpen={this.state.isOpen}
+                position={Position.BOTTOM_LEFT}
+                {...popoverProps}
+                className={classNames(listProps.className, popoverProps.className)}
+                onInteraction={this.handlePopoverInteraction}
+                popoverClassName={classNames("pt-multi-select-popover", popoverProps.popoverClassName)}
+                popoverDidOpen={this.handlePopoverDidOpen}
+            >
+                <div
+                    onKeyDown={this.handleKeyDown}
+                    onKeyUp={this.state.isOpen && handleKeyUp}
+                >
+                    <TagInput
+                        inputProps={defaultInputProps}
```

Comment:
oof this'll get clobbered if the user defines even one custom `inputProp`... gonna have to spread `tagInputProps.inputProps` into `defaultProps`.